File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/AnonymousDeclarationPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.ConverterContext
import io.github.sgrishchenko.karakum.extension.ConverterPlugin
import io.github.sgrishchenko.karakum.extension.DerivedFile
import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.structure.derived.DerivedDeclaration
import io.github.sgrishchenko.karakum.structure.derived.generateDerivedDeclarations
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import js.array.ReadonlyArray
import js.objects.JsPlainObject
import typescript.Node

external interface AnonymousDeclarationContext : ConverterContext {
    fun resolveName(node: Node): String
}

external interface AnonymousDeclaration

@PublishedApi
@JsPlainObject
internal external interface AnonymousDeclarationRenderResult {
    val name: String
    val declaration: String
    val reference: String
}

@Suppress("NOTHING_TO_INLINE")
inline fun AnonymousDeclaration(result: String) =
    result.unsafeCast<AnonymousDeclaration>()

@Suppress("NOTHING_TO_INLINE")
inline fun AnonymousDeclaration(
    name: String,
    declaration: String,
    reference: String,
) =
    AnonymousDeclarationRenderResult(
        name = name,
        declaration = declaration,
        reference = reference,
    ).unsafeCast<AnonymousDeclaration>()

typealias AnonymousDeclarationRender = (
    node: Node,
    context: AnonymousDeclarationContext,
    render: Render<Node>,
) -> AnonymousDeclaration?

class AnonymousDeclarationPlugin(
    private val anonymousDeclarationRender: AnonymousDeclarationRender
) : ConverterPlugin<Node> {
    private val generated = mutableMapOf<Node, DerivedDeclaration>()

    override fun setup(context: ConverterContext) = Unit

    override fun traverse(node: Node, context: ConverterContext) = Unit

    override fun render(node: Node, context: ConverterContext, next: Render<Node>): String? {
        val nameResolverService = context.lookupService<NameResolverService>(nameResolverServiceKey)
        if (nameResolverService == null) error("AnonymousDeclarationPlugin can't work without NameResolverService")

        val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
        if (typeScriptService == null) error("AnonymousDeclarationPlugin can't work without TypeScriptService")

        val annotationService = context.lookupService<AnnotationService>(annotationServiceKey)
        if (annotationService == null) error("AnonymousDeclarationPlugin can't work without AnnotationService")

        val annotations = annotationService.resolveAnonymousAnnotations(node, context)

        val anonymousDeclarationContext = object : AnonymousDeclarationContext, ConverterContext by context {
            override fun resolveName(node: Node) = nameResolverService.resolveName(node, context)
        }

        val result = anonymousDeclarationRender(
            node,
            anonymousDeclarationContext,
            next
        )

        if (result == null) return result
        if (jsTypeOf(result) == "string") return result.toString()

        val renderResult = result.unsafeCast<AnonymousDeclarationRenderResult>()
        val name = renderResult.name
        val declaration = renderResult.declaration
        val reference = renderResult.reference

        val sourceFileName = node.getSourceFileOrNull()?.fileName ?: "generated.d.ts"
        val namespace = typeScriptService.findClosestNamespace(node)

        generated[node] = DerivedDeclaration(
            sourceFileName = sourceFileName,
            namespace = namespace,
            fileName = "${name}.kt",
            body = (annotations + declaration).joinToString(separator = "\n"),
        )

        return reference;
    }

    override fun generate(context: ConverterContext, render: Render<Node>): ReadonlyArray<DerivedFile> {
        return generateDerivedDeclarations(generated.values.toTypedArray(), context)
    }
}

fun createAnonymousDeclarationPlugin(
    render: AnonymousDeclarationRender,
): ConverterPlugin<Node> {
    return AnonymousDeclarationPlugin(render)
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/AnonymousDeclarationPlugin.ts
import ts, {Node} from "typescript";
import {ConverterPlugin} from "../plugin.js";
import {ConverterContext} from "../context.js";
import {Render} from "../render.js";
import {NameResolverService, nameResolverServiceKey} from "./NameResolverPlugin.js";
import {generateDerivedDeclarations} from "../../structure/derived/generateDerivedDeclarations.js";
import {DerivedDeclaration} from "../../structure/derived/derivedDeclaration.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {DerivedFile} from "../generated.js";
import {AnnotationService, annotationServiceKey} from "./AnnotationPlugin.js";

export interface AnonymousDeclarationContext extends ConverterContext {
    resolveName(node: Node): string
}

export interface AnonymousDeclarationRenderResult {
    name: string
    declaration: string
    reference: string
}

export type AnonymousDeclarationRender = (
    node: ts.Node,
    context: AnonymousDeclarationContext,
    render: Render,
) => AnonymousDeclarationRenderResult | string | null

class AnonymousDeclarationPlugin<TNode extends Node = Node> implements ConverterPlugin {
    private readonly generated: Map<Node, DerivedDeclaration> = new Map()

    constructor(
        private anonymousDeclarationRender: AnonymousDeclarationRender,
    ) {
    }

    setup(context: ConverterContext): void {
    }

    traverse(node: ts.Node, context: ConverterContext): void {
    }

    render(node: ts.Node, context: ConverterContext, next: Render): string | null {
        const nameResolverService = context.lookupService<NameResolverService>(nameResolverServiceKey)
        if (nameResolverService === undefined) throw new Error("AnonymousDeclarationPlugin can't work without NameResolverService")

        const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
        if (typeScriptService === undefined) throw new Error("AnonymousDeclarationPlugin can't work without TypeScriptService")

        const annotationService = context.lookupService<AnnotationService>(annotationServiceKey)
        if (annotationService === undefined) throw new Error("AnonymousDeclarationPlugin can't work without AnnotationService")

        const resolveName = (node: TNode) => nameResolverService.resolveName(node, context)
        const annotations = annotationService.resolveAnonymousAnnotations(node, context)

        const anonymousDeclarationContext = {
            ...context,
            resolveName,
        }

        const result = this.anonymousDeclarationRender(
            node,
            anonymousDeclarationContext,
            next
        )

        if (result === null || typeof result === "string") return result

        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
        const namespace = typeScriptService.findClosest(node, ts.isModuleDeclaration)

        const {name, declaration, reference} = result

        this.generated.set(node, {
            sourceFileName,
            namespace,
            fileName: `${name}.kt`,
            body: [...annotations, declaration].join("\n"),
        })

        return reference;
    }

    generate(context: ConverterContext): DerivedFile[] {
        return generateDerivedDeclarations(Array.from(this.generated.values()), context)
    }
}

export function createAnonymousDeclarationPlugin<TNode extends Node = Node>(
    render: AnonymousDeclarationRender,
): ConverterPlugin {
    return new AnonymousDeclarationPlugin<TNode>(render)
}
